storage/innobase/buf/buf0buf.cc

```diff
@@ -742,6 +742,211 @@ bool buf_page_is_corrupted(bool check_lsn, const byte *read_buf,
+  int m_pipe[2];
+
+  std::thread m_thd;
+  /* mem pressure garbarge collection resticted to interval */
+  static const ulonglong max_frequency_ms= 60*1000000;
+public:
+  mem_pressure() : m_num_fds(0), m_pipe{-1, -1} {}
+
+  bool setup()
+  {
+    static_assert(array_elements(m_fds) == (array_elements(m_triggers) + 1),
+                  "insufficient fds");
+    std::string memcgroup{"/sys/fs/cgroup"};
+    std::string cgroup;
+    std::ifstream selfcgroup("/proc/self/cgroup");
+    std::getline(selfcgroup, cgroup, '\n');
+
+    cgroup.erase(0, 3); // Remove "0::"
+    memcgroup+= cgroup + "/memory.pressure";
+
+    m_num_fds= 0;
+    for (auto trig= std::begin(m_triggers); trig!= std::end(m_triggers); ++trig)
+    {
+      if ((m_fds[m_num_fds].fd=
+             my_open(memcgroup.c_str(), O_RDWR | O_NONBLOCK, MYF(0))) < 0)
+      {
+        switch (errno)
+        {
+        case EPERM:
+          sql_print_information("Failed to initialize memory pressure EPERM, file permissions or CAP_SYS_RESOURCE needed");
+          break;
+        default:
+          sql_print_information("Failed to initialize memory pressure: %s", strerror(errno));
+        }
+        shutdown();
+        return false;
+      }
+      ssize_t slen= strlen((const char *)*trig);
+      if (write(m_fds[m_num_fds].fd, *trig, slen) < slen)
+      {
+        sql_print_warning("Failed create trigger for memory pressure \"%s\"", *trig);
+        my_close(m_fds[m_num_fds].fd, MYF(MY_WME));
+        continue;
+      }
+      m_fds[m_num_fds].events= POLLPRI;
+      m_num_fds++;
+    }
+    if (m_num_fds < 1)
+      return false;
+
+    if (pipe2(m_pipe, O_CLOEXEC | O_DIRECT))
+    {
+      sql_print_warning("No memory pressure - can't create pipe");
+      shutdown();
+      return false;
+    }
+    my_register_filename(m_pipe[0], "mem_pressure_read", FILE_BY_DUP, 0, MYF(0));
+    my_register_filename(m_pipe[1], "mem_pressure_write", FILE_BY_DUP, 0, MYF(0));
+    m_fds[m_num_fds].fd= m_pipe[0];
+    m_fds[m_num_fds].events= POLLIN;
+    m_num_fds++;
+    m_thd= std::thread(pressure_routine, this);
+    return true;
+  }
+
+  void shutdown()
+  {
+    while (m_num_fds)
+    {
+      m_num_fds--;
+      my_close(m_fds[m_num_fds].fd, MYF(MY_WME));
+    }
+    if (m_pipe[1] >= 0)
+      my_close(m_pipe[1], MYF(MY_WME));
+  }
+
+  static void pressure_routine(mem_pressure *m)
+  {
+    if (my_thread_init())
+      return;
+#ifdef UNIV_PFS_THREAD
+    pfs_register_thread(mem_pressure_thread_key);
+#endif
+
+    ulonglong last= microsecond_interval_timer() - max_frequency_ms;
+    while (1)
+    {
+      int retval= poll(&m->m_fds[0], m->m_num_fds, -1);
+      if (retval == EINTR)
+        continue;
+      for (nfds_t i= 0; i < m->m_num_fds; ++i)
+      {
+        if (m->m_fds[i].revents & POLLPRI)
+        {
+          ulonglong now= microsecond_interval_timer();
+          if ((now - last) > max_frequency_ms)
+          {
+            last= now;
+            garbage_collect();
```

Comment:
Do we really need the separate `garbage_collect()` function? We could just call `buf_pool.garbage_collect()` and output the message here.